Add option for 'use expression body for single line'

CyrusNajmabadi opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/15414

CyrusNajmabadi commented:
Feature looks like this:

![image](https://cloud.githubusercontent.com/assets/4564579/23837125/a676ad5a-0740-11e7-80ed-97e59cb9f1a0.png)

CyrusNajmabadi commented:
Tagging @dotnet/roslyn-ide This makes the "use expression body" code style more flexible and Allows users to opt into the following on a per-construct basis:

Use expression body for XXX - Never, When possible, When on single line.

With this, a user can choose to not get suggestions to use expression bodies for members that actually contain a large and complex expression.  

Note: the majority of complexity in this PR is around presenting the options to the user and preserving compatibility with our old option.  Specifically, we used to only offer two options 'yes/no' (meaning 'expression bodies are preferred' and 'expression bodies are not preferred').  I had to include code to map those old values to the new ones so that existing editor_config and vs_roaming_settings woudl be understood.

CyrusNajmabadi commented:
Ping, i really need eyes on this.

CyrusNajmabadi commented:
Tagging @dotnet/roslyn-ide Need eyes plz.

Pilchie commented on src/Workspaces/Core/Portable/CodeStyle/CodeStyleOption.cs:
```diff
@@ -77,37 +107,39 @@ public static CodeStyleOption<T> FromXElement(XElement element)
 
-            return ((IEquatable<CodeStyleOption<T>>)this).Equals(codeStyle);
+            var enumValues = (T[])Enum.GetValues(typeof(T));
+            return b ? enumValues[0] : enumValues[1];
```
This is a subtle bit of contract right here...

CyrusNajmabadi replied:
You're write.  I will both try to document this heavily and add tests.

CyrusNajmabadi replied:
I've also made it so that we just go and find the item with the 0 and 1 value instead of assumign they're the 0th and 1st in that array.

CyrusNajmabadi replied:
Fixed.

jmarolf commented on src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs:
```diff
@@ -190,6 +177,24 @@ public bool TryFetch(OptionKey optionKey, out object value)
             return true;
         }
 
+        private bool DeserializeCodeStyleOption<T>(ref object value)
```
>ref [](start = 51, length = 3)

Why `ref` instead of `out`?

CyrusNajmabadi replied:
Because the value is in/out.  i.e. we get it as a string, but we may write back out to is as a CodeStyleOption.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs:
```diff
@@ -190,6 +177,24 @@ public bool TryFetch(OptionKey optionKey, out object value)
+                    value = CodeStyleOption<T>.FromXElement(XElement.Parse(serializedValue));
+                    return true;
+                }
+                catch (Exception)
```
So for the older roaming options that were true/false and are now a tri-state, should we have some affordance here to convert the older values to the newer values?

CyrusNajmabadi replied:
We do.  ```CodeStyleOption<TEnum>``` Will migrate ```CodeStyleOption<bool>``` so that 'false' is mapped to the 0 value of TEnum and true is mapped to the 1 value.   This likely needs a lot of documentation (and tests to validate that this works properly.

CyrusNajmabadi replied:
Docs and tests added.

jasonmalinowski commented on src/VisualStudio/Core/Impl/Options/Style/BooleanCodeStyleOptionViewModel.cs:
```diff
@@ -58,7 +58,7 @@ public override NotificationOptionViewModel SelectedNotificationPreference
             List<CodeStylePreference> preferences = null,
             List<NotificationOptionViewModel> notificationPreferences = null)
-            : base(option, description, truePreview, falsePreview, info, options, groupName, preferences, notificationPreferences)
+            : base(option, description, new[] { truePreview, falsePreview }, info, options, groupName, preferences, notificationPreferences)
```
So are we passing data down to the base class but the base class can't do anything without calling GetPreview? Why not just hold the data at this class's level?

CyrusNajmabadi replied:
This code all sucks.  I will fix up later today and will send out again.

jasonmalinowski commented on src/VisualStudio/Core/Impl/Options/Style/EnumCodeStyleOptionViewModel.cs:
```diff
@@ -0,0 +1,106 @@
+        where T : struct
+    {
+        private static readonly ImmutableArray<T> s_enumValues =
+            ImmutableArray.Create((T[])Enum.GetValues(typeof(T)));
```
Is this guaranteed to be stable? If somebody were to reorder names but keep the original numerical values, what happens?

CyrusNajmabadi replied:
I have checked, and hte answer is yes.  But, that said, this just seems britle.  So i will make passing the values explicit.

jasonmalinowski commented on src/VisualStudio/Core/Impl/Options/Style/EnumCodeStyleOptionViewModel.cs:
```diff
@@ -0,0 +1,106 @@
+        public override string GetPreview()
+        {
+            var index = Preferences.IndexOf(SelectedPreference);
+            return Previews[index];
```
Ditto: why are we passing Previews down to the base type when the interpretation is at the derived type's level?

CyrusNajmabadi replied:
This code sucks.  I will fix it up later and resend.

jasonmalinowski commented on src/Workspaces/CSharp/Portable/Extensions/BlockSyntaxExtensions.cs:
```diff
@@ -38,6 +44,18 @@ internal static class BlockSyntaxExtensions
+                return true;
+            }
+
+            Debug.Assert(preference == ExpressionBodyPreference.WhenOnSingleLine);
```
throw UnexpectedValueException()?

CyrusNajmabadi replied:
Changed to contract.

jasonmalinowski commented on src/Workspaces/Core/Portable/CodeStyle/ExpressionBodyPreference.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    /// <summary>
```
<remarks>, not summary. This isn't a useful description in QuickInfo.

CyrusNajmabadi replied:
Fixed

jasonmalinowski commented on src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs:
```diff
@@ -95,10 +99,144 @@ protected static void FreeAliasMap(Dictionary<string, string> aliasMap)
         protected static Dictionary<string, string> AllocateAliasMap()
+            => s_aliasMapPool.Allocate();
+
+        public bool IsOnSingleLine(SyntaxNode node, bool fullSpan)
```
Why isn't this just asking for the line numbers for the start and end of the node's span, and comparing?

CyrusNajmabadi replied:
How would it get line numbers for an arbitrary node?  The node may be synthetic, in which case is does not belong to a tree.  This would cause us to have to allocate a tree, and then create an entire source text to answer that question.

jmarolf commented on src/Workspaces/Core/Portable/CodeStyle/CodeStyleOption.cs:
```diff
@@ -29,10 +35,34 @@ public CodeStyleOption(T value, NotificationOption notification)
 
+        private object GetValueForSerialization()
+        {
+            if (typeof(T) == typeof(bool))
```
>typeof(T) == typeof(bool) [](start = 16, length = 25)

does pattern matching work in this case?

CyrusNajmabadi replied:
No.  These are not variables that are being checked.  and "typeof(T)" would only pattern match against System.Type.

jmarolf commented on src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs:
```diff
@@ -787,7 +787,7 @@ private string GetXmlOption(PerLanguageOption<CodeStyleOption<bool>> option)
-        private void SetXmlOption(Option<CodeStyleOption<bool>> option, string value)
+        private void SetXmlOption<T>(Option<CodeStyleOption<T>> option, string value)
         {
             var convertedValue = CodeStyleOption<bool>.FromXElement(XElement.Parse(value));
```
>bool [](start = 49, length = 4)

Should this be `T`?

CyrusNajmabadi replied:
Good catch!

CyrusNajmabadi replied:
Fixed

jmarolf commented:
>         public string Style_PreferExpressionBodiedConstructors

How will this work when the user roams settings and we try and parse `CodeStyleOption<bool>` as a `CodeStyleOption<ExpressionBodyPreference>`?  Seems like we need the same logic here as we have for editorconfig.  We should first try to parse the value as bool and if that fails attempt to parse it as ExpressionBodyPreference

---
Refers to: src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs:660 in 237803a. [](commit_id = 237803a7ea67fe010ff902150aa6e94809802ad4, deletion_comment = False)

CyrusNajmabadi commented:
@jmarolf AutomationObject will work properly.  It just calls into ```CodeStyleOption<T>.FromXElement(XElement.Parse(value));``` and that codepath knows how to migrate boolean options forward to enum options.

CyrusNajmabadi commented:
@Pilchie @jmarolf Please take another look.  I've tried to address your core concerns.

@jasonmalinowski I have not addressed your concerns about the StyleOptionViewModel, but i will later today.  It definitely can be cleaned up much more, i just didn't view that as a core concern when making the overall change.  Sorry about that.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/Options/Formatting/StyleViewModel.cs:
```diff
@@ -671,49 +671,85 @@ internal StyleViewModel(OptionSet optionSet, IServiceProvider serviceProvider) :
             };
 
-            var codeBlockPreferences = new List<CodeStylePreference>
+            var expressionBodyPreferences = new List<CodeStylePreference>
```
Fixed.

CyrusNajmabadi commented:
ok @jasonmalinowski I thin i've addressed your concerns as well.

CyrusNajmabadi commented:
@kuhlenh Note that this is a change in one of our existing features and changes how editor_config works for it.  Namely the "csharp_style_expression_bodied_XXX" editor_config options are no longer true/false options.  Instead, you can now specify: 

1. "never"
2. "when_possible"
3. "when_on_single_line"

Note: we still support users who use the 'true'/'false' values for back compat (or for interop with Dev15) purposes.  true/false map to 'when_possible'/'never'.    So dev15 editor_config files will work with future VS versions.

However, the reverse is not true.  A user who picks "when_on_single_line" will not see that value respected in dev15 (because dev15 has no concept of what that option means).  If one of these new values is provided, then dev15 will simply fall back to the default for that option ("true" for properties/accessors, and "false" for method-like declarations).

Pilchie commented on src/Workspaces/Core/Portable/CodeStyle/CodeStyleOption.cs:
```diff
@@ -15,6 +15,12 @@ internal interface ICodeStyleOption
     /// Represents a code style option and an associated notification option.  Supports
     /// being instantiated with T as a <see cref="bool"/> or an <code>enum type</code>.
+    /// 
+    /// CodeStyleOption also has some basic support for migration a <see cref="bool"/> option
```
s/migration/migrating/

Pilchie commented on src/Workspaces/Core/Portable/CodeStyle/CodeStyleOption.cs:
```diff
@@ -118,6 +119,9 @@ public static CodeStyleOption<T> FromXElement(XElement element)
                 case nameof(Boolean):
+                    // Try to map a boolean value.  Either map it to true/false if we're a 
+                    // CodeStyleOption<bool> or map it to the 0 or 1 value for an enum if we're
+                    // a CodeStyleOption<SomeEnumType>.
```
This comment should move (or be duplicated) as a doc comment on `Convert`

jmarolf commented:
@CyrusNajmabadi  My main concern here is still that this will change the settings such that roaming them from a 2017.3 machine to a 2017.0 machine will blow up.

Visual Studio 2017.0 | Visual Studio 2017.3
------------ | -------------
`CodeStyleOption<bool>` | `CodeStyleOption<ExpressionBodyPreference>`

✔️ Visual Studio 2017.0  -> Visual Studio 2017.3

✔️ Visual Studio 2017.0  -> Visual Studio 2017.0 

✔️ Visual Studio 2017.3 -> Visual Studio 2017.3

❌ Visual Studio 2017.3 -> Visual Studio 2017.0

CyrusNajmabadi commented:
@jmarolf Can you take another look?  We shoudl now support "Visual Studio 2017.3 -> Visual Studio 2017.0", and i added tests for this.
